File: classB/17/homework2/task2.hs
import Data.List.Split (splitOn)
import Data.List (transpose)

main :: IO ()
main = do
    csv1 <- readFile "matrix1.csv"
    csv2 <- readFile "matrix2.csv"
    let m1 = csvToMatrix csv1
    let m2 = csvToMatrix csv2
    let result = multiplyMatrix m1 m2
    putStrLn $  if canMultiply m1 m2 then
                    if result == transpose result 
                        then "true"
                        else "false"
                else "undefined"

canMultiply :: Num a => [[a]] -> [[a]] -> Bool
canMultiply m1 m2 = length (head m1) == length m2

multiplyMatrix :: Num a => [[a]] -> [[a]] -> [[a]]
multiplyMatrix m1 m2 = map (\m1row -> map (sum --sum the products
                                            . zipWith (*) --muliply all pairs
                                                m1row) --m1 rows
                                                    (transpose m2)) --transposed m2 rows ( m2 cols )
                                                        m1 --map over m1 rows

csvToMatrix :: String -> [[Double]]
csvToMatrix = (map (map read -- read strings to doubles
                        . (splitOn ","))) -- split values
                            . lines -- split lines

File: classB/02/homework2/task2.rb
require "csv"

matrix1 = CSV.read("./matrix1.csv");
matrix2 = CSV.read("./matrix2.csv");

if matrix1[0].length != matrix2.length
	p 'undefined'
	return
end

rows = matrix1.length
cols = matrix2[0].length

result = Array.new(rows, 0){Array.new(cols, 0)}

for i  in 0..rows - 1
	for j in 0..cols - 1
		for k in 0..matrix1[0].length - 1 
			#print "matrix1: ",matrix1[i][k],"  matrix2:",matrix2[k][j],"\n"
			result[i][j] += matrix1[i][k].to_i * matrix2[k][j].to_i
		end
	end
end

for i  in 0..rows - 1
	for j in 0..cols - 1
		if result[i][j] != ( -1 * result[j][i] )
			p false
			return
		end
	end
end

p true
